book/integration-service.md

```diff
@@ -0,0 +1,271 @@
+
+**IntegrationTestScenario** - The custom resource that describes separate PipelineRuns that are supposed to be run by the snapshot controller. Any number of them can be set by the user and each of them includes information about the Tekton pipeline that will be executed - Tekton bundle, pipeline name and parameters, and the Environment details. The Integration Test Scenarios can also specify the contexts where they will be applied - this can be used to define when the scenario will run (Component or Composite stage - or both). If more than one Integration Test Scenario has the same context, their respective PipelineRuns will be executed in parallel.
+
+**Ephemeral Environment** - The ephemeral copy of an existing environment that’s created by the Integration service for use during a specific Integration pipeline run. After the environment is copied and marked as ephemeral, the Integration service creates a new SnapshotEnvironmentBinding linking the Snapshot under test with the newly created Environment. This triggers a deployment of the Snapshot on the new Environment, handled by the GitOps service. Finally, the Integration service links the credentials for the Environment to the Integration Tekton pipeline run and starts the test run.
+
+
+## Resources 
+Below are the list of CR’s that the integration service is responsible for interacting and performing operations:
+
+### CREATE
+
+| Custom Resource | When? | Why? |
+|---|---|---|
+| Snapshot | Post Component build PipelineRun completes.  | Used as the source input for the test gate pipeline |
+| PipelineRuns | Post Component build PipelineRun completes  | To test the Snapshot  |
+| Release | Post PipelineRun if there is a ReleaseLink for the Application, if the auto-release flag is set on the ReleaseLink | To signal to the Release Service to automatically push the Snapshot to the Production Environment |
+
+### READ
+
+| Custom Resources | When? | Why? |
+|---|---|---|
+| Application & Component | Before creating the Snapshots | To know which build.ImageSpecs to use in the Snapshot |
+| IntegrationTestScenario | Before creating the Component/Composite PipelineRun(s) | To get the information for the PipelineRun - which Tekton bundles and Environments will be used |
+| ReleaseLink | Before creating the Release | To signal the Release Service for next environment promotion (prod) |
+
+### UPDATE
+
+| Custom Resource | When? | Why? |
+|---|---|---|
+| SnapshotEnvironmentBinding | Post Test PipelineRun | To promote the Snapshot to the next environment (Stage/Stable) |
+| Snapshot | Post Test PipelineRun | To mark the snapshot validated so that it can be promoted to the next environment |
+| Component | Post Test PipelineRun | To update the Component with the image spec, in order to update the global candidates list |
+
+### WATCH
+
+| Custom Resource | When? | Why? |
```

Comment:
Is there an Environment CR that is being watched? That is the impression that I had from the overview.

Reply:
This part is in flux as the Environment APIs are under design currently and being expanded.
We are currently experimenting with watching `SnapshotEnvironmentBindings` instead of `Environments` to determine if the `Snapshot` was deployed to a given `Environment`, but this is not yet finalized as far as we can tell.

Reply:
ack. I guess just something to keep in mind to come back to.

Reply:
Would it be reasonable to call this out somewhere as a reminder to come back to it ... or a notice to readers about how it is not resolved yet? It might provide some confusions when reading through as is.

Reply:
Done